File: Workshop 4 Elastiske kollisioner-20230210/Ball.py
import matplotlib.pyplot as plt
import numpy as np

class Ball:
    resolution = 100

    def __init__(self, mass, radius, velocity, position, color, name, ax):
        self.mass = mass
        self.radius = radius
        self.velocity = velocity
        self.position = position
        self.color = color
        circle = self.make_circle()
        self.ball = plt.Polygon(circle, fill=self.color, edgecolor='k')
        if ax: ax.add_patch(self.ball)
        self.text = plt.text(
            self.position[0], self.position[1], name,
            horizontalalignment='center', verticalalignment='center',)

    def update_position(self, t):
        # Update circle with Python matplotlib: self.ball.center = self.position

        self.position = self.get_position_at_time_step(t).reshape((2, 1))
        circle = self.make_circle()
        self.ball.set_xy(circle)
        self.text.set_position(self.position)

    def get_position_at_time_step(self, dt):
        position = self.position.reshape((2, 1)) + dt * self.velocity.reshape((2, 1))
        return position.reshape((2, ))

    def make_circle(self):
        # Create circle in Python: matplotlib.patches.Circle(position, radius, color)

        t = np.linspace(0, 2 * np.pi, self.resolution)
        x_points = np.add(self.position[0], np.multiply(self.radius, np.cos(t)))
        y_points = self.position[1] + self.radius * np.sin(t)
        circle = np.stack((x_points, y_points), axis=1)
        return circle

File: Workshop 4 Elastiske kollisioner-20230210/Matlab/Ball.m
classdef Ball<handle
    %Klasse der beskriver boldene i vores model.
    
    properties
        mass
        radius
        velocity
        position
        ball
        color
        resolution=100;
        text
        velocity_vector
    end
    
    methods
        function obj = Ball(mass,radius,velocity,position,color,name)
            %Genererer en bold med given masse, radius, hastighed og
            %position.
            if nargin>0
                obj.mass=mass;
                obj.radius=radius;
                obj.velocity=velocity;
                obj.position=position;
                obj.color=color;
                circle=obj.make_circle();
                obj.ball=patch(circle(1,:),circle(2,:),obj.color);
                hold on
                obj.text=text(obj.position(1),obj.position(2),name);
            end
        end
        
        function obj = update_position(obj,t)
            obj.position=obj.get_position_at_time_step(t);
            circle=obj.make_circle();
            set(obj.ball,'XData', circle(1,:),'YData',circle(2,:));
            set(obj.text,'Position',obj.position');
        end
        function position=get_position_at_time_step(obj,dt)
            position=obj.position+dt*obj.velocity;
        end
        
        function circle= make_circle(obj)
            t=linspace(0,2*pi,obj.resolution);
            circle=[obj.position(1)+obj.radius*cos(t);obj.position(2)+obj.radius*sin(t)];
        end
    end
end

